File: masterClass/challenges/isSubsequence.js
/* function takes two strings and checks whether the characters in the first string form a subsequence of the characters in the second string.

isSubsequence('hello', 'hello world') -> true
isSubsequence('sing', 'string') -> true
isSubsequence('abc', 'abracadabra') -> true
isSubsequence('abc', 'acb') -> false

solution must have at least:
Time O(n + m)
Space O(1)

*/

const isSubsequence = function (substring, string) {
  let substrPointer = 0;
  let strPointer = 0;
  while (substrPointer < substring.length) {
    if (substring[substrPointer] === string[strPointer]) {
      substrPointer++;
      strPointer++;
    } else if (strPointer >= string.length) {
      return false;
    } else {
      strPointer++;
    }
  }
  return true;
};

isSubsequence('abc', 'acb');

File: masterClass/py/problem_patterns/challenges/is_subsequence.py
def is_subsequence(str1, str2):
    """
    Check if the characters in the 1st string form a substring of the second.
    I.e., do the chars from 1st string all appear in 2nd string, without their order changing
    return bool
    """
    ptr1 = 0
    ptr2 = 0
    while ptr1 < len(str1):
        while str2[ptr2] != str1[ptr1]:
            if ptr2 + 1 == len(str2):
                return False
            ptr2 += 1
        ptr1 += 1
    return True

print(is_subsequence("hello", "helloworld")) # true
print(is_subsequence("sing", "sting")) # true
print(is_subsequence("abc", "abracadabra")) # true
print(is_subsequence("abc", "acb")) # false